src/python/WMComponent/WorkflowUpdater/WorkflowUpdaterPoller.py

```diff
@@ -0,0 +1,170 @@
+            logging.exception(msg)
+            raise WorkflowUpdaterException(msg) from None
+
+    def getPileupDocs(self):
+        """
+        Fetch all pileup documents from MSPileup and preprocess the data.
+        :return: a list of dictionaries in the following format:
+          {"pileupName": string with pileup name,
+           "customName": string with custom pileup name - if any,
+           "rses": list of RSE names,
+           "blocks": list of block names}
+        """
+        mgr = RequestHandler()
+        headers = {'Content-Type': 'application/json'}
+        data = mgr.getdata(self.msPileupUrl, params={}, headers=headers, verb='POST',
+                           ckey=self.userKey, cert=self.userCert, encode=True, decode=True)
+        if data and data.get("result", []):
+            if "error" in data["result"][0]:
+                msg = f"Failed to retrieve MSPileup documents. Error: {data}"
+                raise WorkflowUpdaterException(msg)
+
+        logging.info("A total of %d pileup documents have been retrieved.", len(data["result"]))
+        pileupMap = []
+        for puItem in data["result"]:
+            logging.info("Pileup: %s, custom name: %s, expected at: %s, but currently available at: %s",
+                         puItem['pileupName'], puItem['customName'],
+                         puItem['expectedRSEs'], puItem['currentRSEs'])
+            thisPU = {"pileupName": puItem['pileupName'],
+                      "customName": puItem['customName'],
+                      "rses": puItem['currentRSEs'],
+                      "blocks": []}
+            pileupMap.append(thisPU)
+        return pileupMap
+
+    def findWflowsWithPileup(self, listSpecs):
+        """
+        Given a list of workflow names and their respective specs, load each
+        one of them and filter out those that don't require any pileup dataset.
+        :param listSpecs: a list of dictionary with workflow name and spec path
+        :return: a list with the workflow names that require pileup
+        """
+        wflowsWithPU = []
+        for wflowSpec in listSpecs:
+            try:
+                workloadHelper = WMWorkloadHelper()
+                workloadHelper.load(wflowSpec['spec'])
+                pileupSpecs = workloadHelper.listPileupDatasets()
+                if pileupSpecs:
+                    logging.info("Workflow: %s requires pileup dataset(s): %s",
+                                 wflowSpec['name'], pileupSpecs.values())
+                    wflowsWithPU.append(wflowSpec['name'])
+                else:
+                    logging.info("Workflow: %s does not require any pileup", wflowSpec['name'])
+            except Exception as ex:
+                msg = f"Failed to load spec file for: {wflowSpec['spec']}. Details: {str(ex)}"
+                logging.error(msg)
+        return wflowsWithPU
+
+    def findRucioBlocks(self, uniquePUList, pileupMap):
```

Comment:
Given that this component will have a polling cycle of hours (for the moment set to 3 times a day), and that we normally have around 100 pileup datasets in the system, I decided to trade off wallclock time by readability/simplicity/maintainability. I always try to balance those, and IMO parallelism/concurrency is not going to be critical for this component.

However, given that we are already using pycurl_manager in this module, I could replace this code by a concurrent version to boost this component's performance, but then we need to manage our own rucio token and add extra code for error handling. Let me have a quick look into this.

Reply:
Alan, I hear this argument over and over, and in my view it is not valid. Please don't take it as a criticism.But sooner or later we'll face slowness problem. It may take year(s) to see it but since we have this pattern n many places eventually we'll see its effect. My point was that for new code it is better to try out modern way, i.e. introduce concurrency, of writing the code, either via asyncio or via pycurl. And, doing that I hope that developers will start seeing why technology matter. Yes, it will require additional error handling, yes it will require code-refactoring, and indeed it may be cumbersome in Python, etc. But I doubt that we should write code applying old patterns when issue we are trying to solve can be solved much faster.

Reply:
If you go out for a walk of 1km, would you bring 1 gallon of water as you do for a walk of 100km?

I think it is always important to answer:
a) what is the scenario that I want to cover
b) a projection of this scenario 5 years ahead.
unless there is a good motivation to think that performance will be needed, I tend to always select a more maintainable approach.

I will refactor this function though and make it concurrent.

Reply:
Alan, you are missing one important item in your logic:
c) in which environment I run my function, and how this environment behave under (a) and (b)

I saw this many, many times in different places (WMCore, Unified, Tier0, DBS, Phedex, WMArchive, DAS, etc), and different places contribute differently, You are not running a single function once (even if you doing poling logic), but you are running very complex eco-system (services) which have many moving parts and if you do not taking care of your parts your car will break eventually.

And, thanks for consideration to implement concurrent logic.

Reply:
Item c) is factorized in items a) and b).

After looking closer at this code and almost converging with a candidate concurrent implementation, it would add around 100 lines or more of code to have this support in a concurrent manner, either relying on asyncio or on the pycurl module.

With the pycurl module, we can reuse most of the code from:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/Tools/PycurlRucio.py

however, I definitely do not want to have WMAgent code depending on MicroService. That means that I either have to duplicate code; or refactor/move PycurlRucio under a common area (likely WMCore/Services/Rucio).

Having said that, I decided not to change this code for the moment. If you are really strong about this, I invite you to make another issue and address it in the future.

Reply:
Alan, since we are debating this issue I suggest to open new META issue to address concurrency in WMCore/WMAgent (as do solve it in some part of MS). We have a repeated pattern when some APIs need to process identical set of calls to Rucio or DBS, etc. And this is another example. If you'll de-cipher what rucio python client is doing and find its actual REST API call to Rucio server then you don't need 100 lines of code and neither need RucioClient module from MS services. Once you know Rucio REST API you can easily use pycurl module `multi_getdata` function (used in many places in MS codebase) to get what you want. Said that, DAS is doing this forever, and I always argue that in some cases usage of Python clients is not beneficial and we need to rely on bare REST calls. Anyway, I wanted to raise again that it is potentially an issue within WM codebase and proper attention (via META issue) may be required to adopt concurrency pattern.

Reply:
Here we go: https://github.com/dmwm/WMCore/issues/11800
Feel free to update it with any content that you feel relevant.